Document creating playbooks with Project Ansible Module Collections (#3844)
theforeman/foreman-documentation: guides/common/modules/ref_example-playbooks-based-on-modules-from-project-ansible-collection.adoc

[id="example-playbooks-based-on-modules-from-{project-context}-ansible-collection"]
= Example playbooks based on modules from {Project} Ansible Collection

All playbooks based on modules from {Project} Ansible Collection must include parameters detailing how to connect to the {Project} API.
The following examples use Ansible vault and module defaults group to provide these parameters, and they authenticate using a username and password.
For more information, see xref:creating-a-playbook-with-modules-from-{project-context}-ansible-collection[].

> .Example Ansible playbook: Enable repositories and create a content view

Reviewer: This playbook assumes you have uploaded a manifest and thus have access to the CDN. We should point that out.

Author: Added.

        organization: "Default Organization"
        product: "Red Hat Enterprise Linux for x86_64"
        repositories:
        - releasever: "9"
        state: enabled
    - name: Ensure RHEL 9 AppStream repositories are enabled
      {ansible-namespace}.repository_set:
        name: "Red Hat Enterprise Linux 9 for x86_64 - AppStream (RPMs)"
        organization: "Default Organization"
        product: "Red Hat Enterprise Linux for x86_64"
        repositories:
        - releasever: "9"
        state: enabled
    - name: Ensure a content view for RHEL 9 repositories exists
      {ansible-namespace}.content_view:
        name: "RHEL 9 content view"
        organization: "Default Organization"
        repositories:
          - name: "Red Hat Enterprise Linux 9 for x86_64 - BaseOS RPMs 9"
            product: "Red Hat Enterprise Linux for x86_64"
          - name: "Red Hat Enterprise Linux 9 for x86_64 - AppStream RPMs 9"
            product: "Red Hat Enterprise Linux for x86_64"
----

For more information, see the Ansible module documentation with the following commands:

* `ansible-doc {ansible-namespace}.repository_sync`
* `ansible-doc {ansible-namespace}.content_view`
====

> .Example Ansible playbook: Synchronize repositories and publish a content view

Reviewer: This playbook assumes that the previous (or similar) playbook was successfully ran, we should point that out.

Author: Added.

[id="example-playbooks-based-on-modules-from-{project-context}-ansible-collection"]
= Example playbooks based on modules from {Project} Ansible Collection

All playbooks based on modules from {Project} Ansible Collection must include parameters detailing how to connect to the {Project} API.
The following examples use Ansible vault and module defaults group to provide these parameters, and they authenticate using a username and password.
For more information, see xref:creating-a-playbook-with-modules-from-{project-context}-ansible-collection[].

[id="example-ansible-playbook-enable-repositories-and-create-a-content-view"]
.Example Ansible playbook: Enable repositories and create a content view
====
This example playbook uses the following modules:

* `{ansible-namespace}.repository_set`
* `{ansible-namespace}.content_view`

The playbook ensures repositories are enabled and a content view that contains these repositories exists.

> Before you run this playbook, ensure that you have uploaded a manifest and can access the Red{nbsp}Hat CDN server.

Reviewer: ```suggestion
Before you run this playbook, ensure that you have uploaded a manifest and can access the Red{nbsp}Hat CDN.
```

[id="example-playbooks-based-on-modules-from-{project-context}-ansible-collection"]
= Example playbooks based on modules from {Project} Ansible Collection

All playbooks based on modules from {Project} Ansible Collection must include parameters detailing how to connect to the {Project} API.
The following examples use Ansible vault and module defaults group to provide these parameters, and they authenticate using a username and password.
For more information, see xref:creating-a-playbook-with-modules-from-{project-context}-ansible-collection[].

.Example Ansible playbook: Enable repositories and create a content view
====
This example playbook uses the following modules:

* `{ansible-namespace}.repository_set`

> * `{ansible-namespace}.content_view`

Reviewer: you also use `repository_sync` -- it's technically not necessary for creating the data structures, so I'd probably avoid it.

Author: repository_sync removed from this example

The playbook ensures repositories are enabled and a content view that contains these repositories exists.

[options="nowrap" subs="+quotes,attributes"]
----
- name: Ensure RHEL 10 repositories are enabled and a content view exists
  hosts: localhost
  vars_files:
    - _My_Vault.yml_
  module_defaults:
    group/{ansible-module-defaults-group}:
      username: "{{ _My_Username_ }}"
      password: "{{ _My_Password_ }}"
      server_url: "{{ _My_Server_URL_ }}"
  tasks:
    - name: Ensure RHEL 10 repositories are enabled
      {ansible-namespace}.repository_set:

>         name: "Red Hat Enterprise Linux 10 for x86_64 - BaseOS (RPMs)"

Reviewer: I'd use RHEL9, just for the sake of it already being released :)

Author: Replaced

        organization: "Default Organization"
        product: "Red Hat Enterprise Linux for x86_64"
        repositories:
        - releasever: "10"

>           basearch: "x86_64"

Reviewer: Since RHEL8 the `basearch` is not needed (and actually breaks things sometimes), as the `x86_64` is already in the product/repo name.

Author: Good to know, and `basearch` removed.

Author: I wonder if it means the examples in https://github.com/theforeman/foreman-ansible-modules/blob/develop/plugins/modules/repository_set.py should be updated. They are for RHEL 7 but still. Perhaps it might be a good idea to use a more recent version as an example?

Reviewer: There are rhel8 examples if you scroll down, but indeed they could be more prominent

>     - name: Ensure RHEL 10 repositories are enabled
>       {ansible-namespace}.repository_set:
>         name: "Red Hat Enterprise Linux 10 for x86_64 - BaseOS (RPMs)"
>         organization: "Default Organization"
>         product: "Red Hat Enterprise Linux for x86_64"
>         repositories:
>         - releasever: "10"
>           basearch: "x86_64"
>         state: enabled
>       {ansible-namespace}.repository_set:
>         name: "Red Hat Enterprise Linux 10 for x86_64 - AppStream (RPMs)"
>         organization: "Default Organization"
>         product: "Red Hat Enterprise Linux for x86_64"
>         repositories:
>         - releasever: "10"
>           basearch: "x86_64"
>         state: enabled

Reviewer: ```suggestion
    - name: Ensure RHEL 9 BaseOS repositories are enabled
      {ansible-namespace}.repository_set:
        name: "Red Hat Enterprise Linux 9 for x86_64 - BaseOS (RPMs)"
        organization: "Default Organization"
        product: "Red Hat Enterprise Linux for x86_64"
        repositories:
        - releasever: "9"
        state: enabled
    - name: Ensure RHEL 9 AppStream repositories are enabled
      {ansible-namespace}.repository_set:
        name: "Red Hat Enterprise Linux 9 for x86_64 - AppStream (RPMs)"
        organization: "Default Organization"
        product: "Red Hat Enterprise Linux for x86_64"
        repositories:
        - releasever: "9"
        state: enabled
```

these need to be separate tasks with a name etc. I guess this is the place where you considered doing loops?

>     - name: Sync RHEL 10 repositories
>       {ansible-namespace}.repository_sync:
>         repository: "Red Hat Enterprise Linux 10 for x86_64 - BaseOS RPMs 10"
>         product: "Red Hat Enterprise Linux for x86_64"
>         organization: "Default Organization"
>       {ansible-namespace}.repository_sync:
>         repository: "Red Hat Enterprise Linux 10 for x86_64 - AppStream RPMs 10"
>         product: "Red Hat Enterprise Linux for x86_64"
>         organization: "Default Organization"

Reviewer: ```suggestion
    - name: Sync RHEL repositories
      {ansible-namespace}.repository_sync:
        product: "Red Hat Enterprise Linux for x86_64"
        organization: "Default Organization"
```

Shortcut! You can just name the product, and it will sync all repos of that product :)

        repositories:
          - name: "Red Hat Enterprise Linux 10 for x86_64 - BaseOS RPMs 10"
            product: "Red Hat Enterprise Linux for x86_64"
          - name: "Red Hat Enterprise Linux 10 for x86_64 - AppStream RPMs 10"
            product: "Red Hat Enterprise Linux for x86_64"
----

For more information, see the Ansible module documentation with the following commands:

* `ansible-doc {ansible-namespace}.repository_set`
* `ansible-doc {ansible-namespace}.content_view`
====

.Example Ansible playbook: Synchronize repositories and publish a content view
====
This example playbook uses the following modules:

* `{ansible-namespace}.repository_sync`
* `{ansible-namespace}.content_view_version`

The playbook synchronizes repositories and publishes the content view that includes them.

[options="nowrap" subs="+quotes,attributes"]
----
- name: Ensure RHEL 10 repositories are synced and content view is published
  hosts: localhost
  vars_files:
    - _My_Vault.yml_
  module_defaults:
    group/{ansible-module-defaults-group}:
      username: "{{ _My_Username_ }}"
      password: "{{ _My_Password_ }}"
      server_url: "{{ _My_Server_URL_ }}"
  tasks:

>     - name: Sync RHEL 10 repositories
>       {ansible-namespace}.repository_sync:
>         repository: "Red Hat Enterprise Linux 10 for x86_64 - BaseOS RPMs 10"
>         product: "Red Hat Enterprise Linux for x86_64"
>         organization: "Default Organization"
>       {ansible-namespace}.repository_sync:
>         repository: "Red Hat Enterprise Linux 10 for x86_64 - AppStream RPMs 10"
>         product: "Red Hat Enterprise Linux for x86_64"
>         organization: "Default Organization"

Reviewer: ```suggestion
    - name: Sync RHEL repositories
      {ansible-namespace}.repository_sync:
        product: "Red Hat Enterprise Linux for x86_64"
        organization: "Default Organization"
```

Shortcut again :)

    - name: Publish RHEL 10 content view
      {ansible-namespace}.content_view_version:
        content_view: "RHEL 10 content view"
        organization: "Default Organization"
----

For more information, see the Ansible module documentation with the following commands:

> * `ansible-doc {ansible-namespace}.repository_set`

Reviewer: you're not using that here

Author: Fixed
